crit/explore.go

```diff
@@ -373,3 +373,122 @@ func (c *crit) ExploreRss() ([]*RssMap, error) {
 
 	return rssMaps, nil
 }
+
+// Sk represents the sockets associated with a single process
+type Sk struct {
+	PId     uint32    `json:"pId"`
+	Sockets []*Socket `json:"sockets"`
+}
+
+// Socket represents a single socket
+type Socket struct {
+	Fd       uint32 `json:"fd"`
+	Type     string `json:"type"`
+	Family   string `json:"family,omitempty"`
+	Protocol string `json:"protocol,omitempty"`
+	State    string `json:"state,omitempty"`
+	SrcAddr  string `json:"srcAddr,omitempty"`
+	SrcPort  uint32 `json:"srcPort,omitempty"`
+	DestAddr string `json:"destAddr,omitempty"`
+	DestPort uint32 `json:"destPort,omitempty"`
+	SendBuf  string `json:"sendBuf,omitempty"`
+	RecvBuf  string `json:"recvBuf,omitempty"`
+}
+
+// ExploreSk searches the process tree for sockets
+// and returns a list of PIDs with the associated sockets
+func (c *crit) ExploreSk() ([]*Sk, error) {
+	psTreeImg, err := getImg(filepath.Join(c.inputDirPath, "pstree.img"), &pstree.PstreeEntry{})
+	if err != nil {
+		return nil, err
+	}
+
+	sks := make([]*Sk, 0)
+	for _, entry := range psTreeImg.Entries {
+		process := entry.Message.(*pstree.PstreeEntry)
+		pID := process.GetPid()
+		// Get file with object IDs
+		idsImg, err := getImg(filepath.Join(c.inputDirPath, fmt.Sprintf("ids-%d.img", pID)), &criu_core.TaskKobjIdsEntry{})
+		if err != nil {
+			return nil, err
+		}
+		filesID := idsImg.Entries[0].Message.(*criu_core.TaskKobjIdsEntry).GetFilesId()
+		// Get open file descriptors
+		fdInfoImg, err := getImg(filepath.Join(c.inputDirPath, fmt.Sprintf("fdinfo-%d.img", filesID)), &fdinfo.FdinfoEntry{})
+		if err != nil {
+			return nil, err
+		}
+		skEntry := Sk{PId: pID}
+		for _, fdInfoEntry := range fdInfoImg.Entries {
+			fdInfo := fdInfoEntry.Message.(*fdinfo.FdinfoEntry)
+			socket := Socket{
+				Fd:   fdInfo.GetFd(),
+				Type: fdInfo.GetType().String(),
+			}
+			switch fdInfo.GetType() {
+			case fdinfo.FdTypes_INETSK:
+				file, err := getFile(c.inputDirPath, fdInfo.GetId())
+				if err != nil {
+					return nil, err
+				}
```

Comment:
It might be good to move this code just before the `switch` statement to avoid code duplication.
```Go
file, err := getFile(c.inputDirPath, fdInfo.GetId())
if err != nil {
	return nil, err
}
```